org.eclipse.jdt.ls.tests/src/org/eclipse/jdt/ls/core/internal/handlers/CompletionHandlerTest.java

```diff
@@ -2622,8 +2625,11 @@ public void testLimitCompletionResults() throws Exception {
 	@Test
 	public void testStaticImports2() throws Exception {
 		PreferenceManager.getPrefs(null).setJavaCompletionFavoriteMembers(Collections.emptyList());
-		ICompilationUnit unit = getWorkingCopy("src/test1/B.java",
-		//@formatter:off
+		long timeout = Long.getLong("completion.timeout", 5000);
+		try {
+			System.setProperty("completion.timeout", String.valueOf(60000));
```

Comment:
With the new PR, i also observed the completion performance on type degraded.

Environment:
- Hardware
    * MacBook Pro (13-inch, 2016, Four Thunderbolt 3 Ports)
    * Procesor: 2.9 GHz Dual-Core Intel Core i5
    * Memory: 8 GB 2133 MHz LPDDR3
- Project: https://github.com/spring-projects/spring-petclinic
- VS Code 1.48, only with Language Support for Java enabled.

Old:
![completion2](https://user-images.githubusercontent.com/14052197/90615730-1a488a00-e23f-11ea-8ae1-6610c4a9cc18.gif)

New: 
It takes more than 3s to get the completion of `String`. If you only type `S`, it's slower.
![completion1](https://user-images.githubusercontent.com/14052197/90615718-17e63000-e23f-11ea-9262-deb9f14d39c8.gif)

This issue should be related with the new `Substring/Subword matches for types` https://www.eclipse.org/eclipse/news/4.17/jdt.php#content-assist-substring-types.

It seems it only provides switch to control `subwords match`, we need evaluate more on the new change.

Reply:
@testforstephen can you try disabling subwords match in the JDT preferences, see if it restores the performance?

Reply:
Sure. I tried to turn off both substring and subwords, the performance go back to normal. And the major perf factor is the substring match.

- Turn off substring match with vmArgs `-Djdt.codeCompleteSubstringMatch=false`. We should add this to all *.launch files and the frontend ts code.
- Turn off subwords match with JavaCore.setOptions().
```java
// JavaLanguageServerPlugin.java

public void start(BundleContext bundleContext) throws Exception {
    ...
    Hashtable<String, String> options = new Hashtable<>();
    options.put(AssistOptions.OPTION_SubwordMatch, AssistOptions.DISABLED);
    JavaCore.setOptions(options);
}
```

Reply:
i suggest we keep the max completion timeout to 5s in the test case. This will guard the possible performance issue.

Reply:
options.put(AssistOptions.OPTION_SubwordMatch, AssistOptions.DISABLED) should be added in the preferencemanager

Reply:
> i suggest we keep the max completion timeout to 5s in the test case. This will guard the possible performance issue.

Some tests fail randomly without this option. See https://github.com/eclipse/eclipse.jdt.ls/pull/1411#pullrequestreview-395427762

Reply:
> 

> options.put(AssistOptions.OPTION_SubwordMatch, AssistOptions.DISABLED) 
> should be added in the preferencemanager

Done

Reply:
Please turn off substring match too. It's the main factor to slow down the completion performance. Find the switch in my comment above.

Reply:
It is done. See https://github.com/eclipse/eclipse.jdt.ls/pull/1525/commits/59f3ea04aa516ad37d613ca6707c04421479f634#diff-ce44e5aecff47a5dd96a42b0a961f48cR86

Reply:
No, that line only turns off subword match. The switch for substring match is `System.setProperty(AssistOptions.PROPERTY_SubstringMatch, "false")`

Reply:
Done. See https://github.com/eclipse/eclipse.jdt.ls/pull/1525/commits/f52096613015aae6840cc047e01cf51b93edaf45#diff-55f26b9037b6c92ad4701f61028ae2cdR190

Reply:
Just a heads-up regarding this feature. It might be worth revisiting once we adopt 2021-03 M1.

https://www.eclipse.org/lists/platform-dev/msg02635.html

Reply:
The parallel index search looks promising.

Reply:
I tried to test this and while I think there might be a bit of an improvement using JDT Core 2021-03 M1, it still was much slower than with subword feature completely disabled so will have to look on the upstream bug.